pkg/kubelet/eviction/threshold_notifier_linux.go

```diff
@@ -67,51 +100,105 @@ func NewMemCGThresholdNotifier(path, attribute, threshold, description string, h
+	epfd, err := unix.EpollCreate1(0)
+	if epfd == -1 {
+		glog.Warningf("eviction manager: error creating epoll: %v", err)
+		return
+	}
+	event := unix.EpollEvent{
+		Fd:     int32(n.eventfd),
+		Events: unix.EPOLLIN,
+	}
+	err = unix.EpollCtl(epfd, unix.EPOLL_CTL_ADD, n.eventfd, &event)
+	if err != nil {
+		glog.Warningf("eviction manager: error adding epoll eventfd: %v", err)
+		return
+	}
 	for {
-		buf := make([]byte, 8)
-		_, err := unix.Read(eventfd, buf)
+		event, err := wait(epfd, n.eventfd)
 		if err != nil {
+			glog.Warningf("eviction manager: error while waiting for memcg events: %v", err)
 			return
+		} else if event {
+			// Consume the event from the eventfd
+			buf := make([]byte, 8)
+			_, err := unix.Read(n.eventfd, buf)
+			if err != nil {
+				glog.Warningf("eviction manager: error reading memcg events: %v", err)
+				return
+			}
+			select {
+			case <-n.stop:
+				return
+			case eventCh <- struct{}{}:
+			}
 		}
 
-		select {
-		case eventCh <- struct{}{}:
-		case <-stop.Ch():
+		if _, ok := <-n.stop; !ok {
 			return
 		}
 	}
 }
 
-func (n *memcgThresholdNotifier) Start(stop ThresholdStopCh) {
-	eventCh := make(chan struct{})
-	go getThresholdEvents(n.eventfd, eventCh, stop)
-	for {
-		select {
-		case <-stop.Ch():
-			glog.V(3).Infof("eviction: stopping threshold notifier")
-			unix.Close(n.watchfd)
-			unix.Close(n.controlfd)
-			unix.Close(n.eventfd)
-			return
-		case <-eventCh:
-			glog.V(2).Infof("eviction: threshold crossed")
-			n.handler(n.description)
+// wait waits up to notifierRefreshInterval for an event on the Epoll FD for the
+// eventfd we are concerned about.  It returns an error if one occurrs, and true
+// if the consumer should read from the eventfd.
+func wait(epfd, eventfd int) (bool, error) {
+	events := make([]unix.EpollEvent, 7)
+	timeoutMS := int(notifierRefreshInterval / time.Millisecond)
+	n, err := unix.EpollWait(epfd, events, timeoutMS)
+	if n == -1 {
+		if err == unix.EINTR {
+			// Interrupt, ignore the error
+			return false, nil
 		}
+		return false, err
+	}
+	if n == 0 {
+		// Timeout
+		return false, nil
+	}
+	if n > 6 {
+		return false, fmt.Errorf("epoll_wait returned more events than we know what to do with")
 	}
+	for _, event := range events[:n] {
+		if event.Fd == int32(eventfd) {
+			if event.Events&unix.EPOLLHUP != 0 || event.Events&unix.EPOLLERR != 0 || event.Events&unix.EPOLLIN != 0 {
+				// EPOLLHUP: should not happen, but if it does, treat it as a wakeup.
+
+				// EPOLLERR: If an error is waiting on the file descriptor, we should pretend
+				// something is ready to read, and let unix.Read pick up the error.
+
+				// EPOLLIN: There is data to read.
+				return true, nil
+			}
+		}
+	}
+	// An event occurred that we don't care about.
+	return false, nil
+}
+
+// close closes all file descriptors associated with the cgroup notifier, and ensures all goroutines associated with it exit
+func (n *cgroupNotifier) close() {
```

Comment:
I think you also need to close the `epfd` here?

Reply:
done